### PR TITLE
Introduce `eager` option to function-based modifiers

### PR DESCRIPTION
**Pull request by chriskrycho**
Introduce an options argument to the `modifier()` function used to create function-based modifiers, with a single initial option: `eager`. Passing `{ eager: false }` opts out of the previous default beahvior of eagerly consuming all arguments, so that function-based modifiers can have normal auto-tracking semantics -- that is, only rerunning when the tracked state they use changes. This allows users to opt into the new (`eager: false`) behavior on a per-modifier basis.

There are three changes to accompany the new options object here:

1.  Add overloads to `modifier` for the following cases:

    - the user passes no options (the default up till now)
    - the user passes `{ eager: true }` (matching the previous default)
    - the user passes `{ eager: false }` (the new target behavior)

    These overloads intentionally do *not* support invoking with a "signature" for `{ eager: false }`. If users want to introduce an explicit signature for function-based modifiers, they should first convert to the `{ eager: false }` semantics.

2.  Update the function-based modifier manager internals to support this. As a result, we now need to create a new instance each time, with the corresponding options.

3.  Update the internals of the function-based modifier manager to use a state bucket, aligning it with the internals of the class-based modifier manager and getting rid of the need for `Map`s for keeping track of modifier elements and teardowns.

---

(Docs and deprecation of the legacy behavior will follow!)

### CODE REVIEW

File: addon/-private/functional/modifier.ts [hunk 1]
```diff
@@ -103,12 +163,12 @@ export default function modifier(
   // returns an opaque `Modifier` type so that we can provide a result from this
   // type which is useful to TS-aware tooling (e.g. Glint).
   return setModifierManager(
-    () => FunctionalModifierManager,
+    () => new FunctionalModifierManager(options),
```
Comment (rwjblue) [inline addon/-private/functional/modifier.ts:166]: Instead of making a new instance per `modifier(` call, let's just make two: one for eager and one for lazy; then hand back those singeltons...

**Review by rwjblue — CHANGES_REQUESTED** [review]

File: addon/-private/functional/modifier.ts (hunk 1 above)
Comment (chriskrycho) [reply addon/-private/functional/modifier.ts:166]: Ahhh, thank you! This was the one thing I was thinking about.

File: addon/-private/functional/modifier-manager.ts [hunk 2]
```diff
@@ -1,98 +1,101 @@
...
+  }
+
   createModifier(
     factoryOrClass:
       | Factory<FunctionalModifierDefinition<S>>
       | FunctionalModifierDefinition<S>
-  ): FunctionalModifierDefinition<S> {
-    const Modifier = isFactory(factoryOrClass)
+  ): CreatedState<S> {
+    const instance = isFactory(factoryOrClass)
```
Comment (rwjblue) [inline addon/-private/functional/modifier-manager.ts:53]: minor note here: this ternary should be removed in 4.0 (when we drop < ember-source@3.22 support)

File: addon/-private/functional/modifier-manager.ts [hunk 3]
```diff
@@ -1,98 +1,101 @@
...
-
-    setup(modifier, element, args);
   }
 
-  destroyModifier(modifier: FunctionalModifierDefinition<S>): void {
-    teardown(modifier);
+  destroyModifier(state: InstalledState<S>): void {
+    if (state.teardown) {
+      state.teardown();
+    }
```
Comment (rwjblue) [inline addon/-private/functional/modifier-manager.ts:99]: Pretty syntax?
```suggestion
    state.teardown?()
```

File: addon/-private/functional/modifier.ts [hunk 4]
```diff
@@ -54,14 +60,72 @@ type Teardown = () => unknown;
 export default function modifier<
   E extends Element,
   P extends unknown[],
-  N extends Record<string, unknown>
+  N extends object
```
Comment (rwjblue) [inline addon/-private/functional/modifier.ts:63]: I don't fully grok this change, what does changing to `object` vs `Record<string, unknown>` get us?

File: addon/-private/functional/modifier-manager.ts (hunk 2 above)
Comment (chriskrycho) [reply addon/-private/functional/modifier-manager.ts:53]: Yep, noted the same.

File: addon/-private/functional/modifier-manager.ts (hunk 3 above)
Comment (chriskrycho) [reply addon/-private/functional/modifier-manager.ts:99]: I personally hate reading `state.teardown?.()` 😂

File: addon/-private/functional/modifier.ts [hunk 5]
```diff
@@ -54,14 +60,72 @@ type Teardown = () => unknown;
...
   Element: E;
 }>;
 
+/**
+ * An API for writing simple modifiers.
+ *
+ * This function runs the first time when the element the modifier was applied
+ * to is inserted into the DOM, and it *autotracks* while running. Any values
+ * that it accesses will be tracked, including the arguments it receives, and if
+ * any of them changes, the function will run again.
```
Comment (rwjblue) [inline addon/-private/functional/modifier.ts:77]: Since this is the docs for `eager: true` side, should this also mention that all of the incoming arguments are also eagerly consumed and will force the function to run again?

File: addon/-private/functional/modifier.ts [hunk 6]
```diff
@@ -54,14 +60,72 @@ type Teardown = () => unknown;
...
+  Element: E;
+}>;
+
+/**
+ * An API for writing simple modifiers.
+ *
+ * This function runs the first time when the element the modifier was applied
+ * to is inserted into the DOM, and it *autotracks* while running. Any values
+ * that it accesses will be tracked, including the arguments it receives, and if
+ * any of them changes, the function will run again.
```
Comment (rwjblue) [inline addon/-private/functional/modifier.ts:106]: Maybe reword slightly to ensure that it's clear that _only_ arguments that are "consumed" are included in the "will run again" part.

File: addon/-private/functional/modifier.ts [hunk 7]
```diff
@@ -89,26 +154,28 @@ export default function modifier(
   fn: (
     element: Element,
     positional: unknown[],
-    named: Record<string, undefined>
-  ) => void | Teardown
+    named: object
+  ) => void | Teardown,
+  options?: { eager: boolean }
```
Comment (rwjblue) [inline addon/-private/functional/modifier.ts:159]: Can we default here?
```suggestion
  options?: { eager: boolean } = { eager: true }
```

This makes it a bit clearer what the current default is (to my eyes at least)

File: addon/-private/functional/modifier.ts [hunk 8]
```diff
@@ -89,26 +154,28 @@ export default function modifier(
...
-    Named: Record<string, unknown>;
+    Named: object;
     Positional: unknown[];
   };
 }> {
   // SAFETY: the cast here is a *lie*, but it is a useful one. The actual return
   // type of `setModifierManager` today is `void`; we pretend it actually
   // returns an opaque `Modifier` type so that we can provide a result from this
   // type which is useful to TS-aware tooling (e.g. Glint).
+  const isEager = !options || options?.eager;
```
Comment (rwjblue) [inline addon/-private/functional/modifier.ts:171]: This const should be moved _up_ above the `SAFETY` note just above; the note is talking about the line just below the `const` (the `return`) and having hte const between them confused me initially 😛

File: addon/-private/functional/modifier-manager.ts (hunk 3 above)
Comment (rwjblue) [reply addon/-private/functional/modifier-manager.ts:99]: heheheehe fair enough!

File: addon/-private/functional/modifier.ts (hunk 5 above)
Comment (chriskrycho) [reply addon/-private/functional/modifier.ts:77]: Good call, yes.

File: addon/-private/functional/modifier.ts (hunk 8 above)
Comment (chriskrycho) [reply addon/-private/functional/modifier.ts:171]: Ahhh, heh, whoops, yes. Thank you!

File: addon/-private/functional/modifier.ts (hunk 7 above)
Comment (chriskrycho) [reply addon/-private/functional/modifier.ts:159]: Ah! Also lets us simplify the body a bit, too: it can just become a check on `options.eager` directly.

**Review by rwjblue — APPROVED** [review]
